spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreTemplate.java

```diff
@@ -267,8 +268,21 @@ public long count(Class<?> entityClass) {
 	}
 
 	@Override
-	public <T> Slice<?> queryKeysOrEntitiesSlice(StructuredQuery query, Class<T> entityClass, Pageable pageable) {
-		DatastoreResultsIterable<?> results = queryKeysOrEntities(applyPageable(query, pageable), entityClass);
+	public <T> Slice<Key> queryKeysSlice(KeyQuery query, Class<T> entityClass, Pageable pageable) {
+		return buildSlice(query, pageable, Key.class);
+	}
+
+	@Override
+	public <T> Slice<T> queryEntitiesSlice(StructuredQuery query, Class<T> entityClass, Pageable pageable) {
+		if (query instanceof KeyQuery) {
+			throw new DatastoreDataException("query must be an EntityQuery or a ProjectionEntityQuery");
```

Comment:
Why don't you check for those types in the condition then? What if other query types are added in the future?